### ISSUE TITLE
How to select which variant file to download

### ISSUE THREAD

**Issue by Cyklet** [Cyklet/VidLoader#7]
#EXTM3U
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=2200000
https://d2c8xygjp78w85.cloudfront.net/629/204238-879296-00001865_20200616185455.m3u8
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=910800
https://d2c8xygjp78w85.cloudfront.net/629/204238-879296-00001866_20200616185455.m3u8
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=422400
https://d2c8xygjp78w85.cloudfront.net/629/204238-879296-00001867_20200616185455.m3u8
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=64000
https://d2c8xygjp78w85.cloudfront.net/629/204238-879296-00001868_20200616185455.m3u8

i have 4 links like this. How can i choose to download at the 1st, 2nd or 3rd link?
by [@nxnm](https://github.com/nxnam)

**Comment by Cyklet** [comment]
Good question, I hate to have this kind of answers but, at the moment we can't preselect the video quality, in [apple documentation](https://developer.apple.com/library/archive/documentation/AudioVideo/Conceptual/MediaPlaybackGuide/Contents/Resources/en.lproj/HTTPLiveStreaming/HTTPLiveStreaming.html) is written: `You can use the options dictionary to target a particular variant bit rate or a specific media selection to download. If no options are specified, the highest quality variants of the user’s primary audio and video content are downloaded.` 
At the moment it will download the highest quality that user phone can handle, predefined quality can be a good improvement for download functionality, I will try to add it.

**Comment by nxnam** [comment]
I really wish you could add it soon. If I choose a bit rate, then how should I choose a bit rate to match the video quality I need

**Comment by Cyklet** [comment]
We can see in [AVAssetDownloadURLSession documentation](https://developer.apple.com/documentation/avfoundation/avassetdownloadurlsession/1650938-makeassetdownloadtask) we can send `AVAssetDownloadTaskMinimumRequiredMediaBitrateKey ` that is a simple `NSNumber` that requires simple update on public API, we can also send `AVAssetDownloadTaskMediaSelectionKey` I am not quite sure how selection key works I need to investigate a little bit more.

Regarding quality question I need more time to investigate it, we can try to convert bit rate to quality but if it has different fps the quality will be wrong.
Using my framework multiple times I have observed that some of the master playlists can have `RESOLUTION=1280x720` as information that we can use, then we can select the variant that matches with resolution from the initial configuration (will not work in 100% of the time, like in your case, we don't get resolution information from the master playlist so I would like to avoid this approach)

**Comment by nxnam** [comment]
I got it. thank you 😁

**Comment by Cyklet** [comment]
I have added the possibility to send the [AVAssetDownloadTaskMinimumRequiredMediaBitrateKey](https://developer.apple.com/documentation/avfoundation/avassetdownloadtaskminimumrequiredmediabitratekey) value to AVAssetDownloadURLSession. I have tested it and seems that it is working, I will close this issue. 
Regarding the failed state of the variant link that u have provided more investigation is required. I will leave more description in issue #6.
P.S the latest code updates are on the master soon I will add a new tag, I hope to fix the variant issue
